quickwit/quickwit-common/src/tower/load_shed.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use pin_project::pin_project;
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+use tower::{Layer, Service};
+
+/// Tracks the number of in-flight requests being processed by a service and rejects new incoming
+/// requests if the number of in-flight requests exceeds a specified limit.
+#[derive(Debug)]
+pub struct LoadShed<S> {
+    inner: S,
+    permits: Arc<Semaphore>,
+    permit_opt: Option<OwnedSemaphorePermit>,
+}
+
+impl<S> Clone for LoadShed<S>
+where S: Clone
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            permits: self.permits.clone(),
+            permit_opt: None,
+        }
+    }
+}
+
+pub trait MakeLoadShedError {
+    fn make_load_shed_error() -> Self;
+}
+
+impl<S, R> Service<R> for LoadShed<S>
+where
+    S: Service<R>,
+    S::Error: MakeLoadShedError,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = LoadShedFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.permit_opt.is_none() {
+            if let Ok(permit) = self.permits.clone().try_acquire_owned() {
+                self.permit_opt = Some(permit);
+            } else {
+                return Poll::Ready(Err(S::Error::make_load_shed_error()));
```

Comment:
I am not super familiar with tower services. Are we supposed to attach the waker in `cx` to the acquisition of the semaphore?

Reply:
https://github.com/tower-rs/tower/blob/39adf5c509a1b2141f679654d8317524ca96b58b/tower/src/limit/concurrency/service.rs#L65